Move Stack on to `persistent > 2.13`

See #5983. On Windows, at least, Stack does not build with `nightly-2022-12-22` and a dependency on `persistent-2.14.4.3`. The build error is:

~~~text
[ 31 of 115] Compiling Stack.Storage.Util

src\Stack\Storage\Util.hs:43:9: error:
    • Could not deduce (SafeToInsert record)
        arising from a use of ‘insertMany_’
      from the context: (PersistEntityBackend record
                         ~ BaseBackend backend,
                         PersistField parentid, PersistField value, Ord value,
                         PersistEntity record, MonadIO m, PersistQueryWrite backend)
        bound by the type signature for:
                   updateSet :: forall record backend parentid value (m :: * -> *).
                                (PersistEntityBackend record ~ BaseBackend backend,
                                 PersistField parentid, PersistField value, Ord value,
                                 PersistEntity record, MonadIO m, PersistQueryWrite backend) =>
                                (parentid -> value -> record)
                                -> EntityField record parentid
                                -> parentid
                                -> EntityField record value
                                -> Set value
                                -> Set value
                                -> ReaderT backend m ()
        at src\Stack\Storage\Util.hs:(21,1)-(36,27)
      Possible fix:
        add (SafeToInsert record) to the context of
          the type signature for:
            updateSet :: forall record backend parentid value (m :: * -> *).
                         (PersistEntityBackend record ~ BaseBackend backend,
                          PersistField parentid, PersistField value, Ord value,
                          PersistEntity record, MonadIO m, PersistQueryWrite backend) =>
                         (parentid -> value -> record)
                         -> EntityField record parentid
                         -> parentid
                         -> EntityField record value
                         -> Set value
                         -> Set value
                         -> ReaderT backend m ()
    • In the first argument of ‘($)’, namely ‘insertMany_’
      In a stmt of a 'do' block:
        insertMany_
          $ map (recordCons parentId) $ Set.toList (Set.difference new old)
      In the second argument of ‘($)’, namely
        ‘do deleteWhere
              [parentFieldCons ==. parentId,
               valueFieldCons <-. Set.toList (Set.difference old new)]
            insertMany_
              $ map (recordCons parentId) $ Set.toList (Set.difference new old)’
   |
43 |         insertMany_ $
   |         ^^^^^^^^^^^

[ 32 of 115] Compiling Stack.Types.TemplateName

src\Stack\Storage\Util.hs:71:9: error:
    • Could not deduce (SafeToInsert record)
        arising from a use of ‘insertMany_’
      from the context: (PersistEntityBackend record
                         ~ BaseBackend backend,
                         PersistField parentid, Ord value, PersistEntity record, MonadIO m,
                         PersistQueryWrite backend)
        bound by the type signature for:
                   updateList :: forall record backend parentid value (m :: * -> *).
                                 (PersistEntityBackend record ~ BaseBackend backend,
                                  PersistField parentid, Ord value, PersistEntity record, MonadIO m,
                                  PersistQueryWrite backend) =>
                                 (parentid -> Int -> value -> record)
                                 -> EntityField record parentid
                                 -> parentid
                                 -> EntityField record Int
                                 -> [value]
                                 -> [value]
                                 -> ReaderT backend m ()
        at src\Stack\Storage\Util.hs:(47,1)-(61,27)
      Possible fix:
        add (SafeToInsert record) to the context of
          the type signature for:
            updateList :: forall record backend parentid value (m :: * -> *).
                          (PersistEntityBackend record ~ BaseBackend backend,
                           PersistField parentid, Ord value, PersistEntity record, MonadIO m,
                           PersistQueryWrite backend) =>
                          (parentid -> Int -> value -> record)
                          -> EntityField record parentid
                          -> parentid
                          -> EntityField record Int
                          -> [value]
                          -> [value]
                          -> ReaderT backend m ()
    • In the first argument of ‘($)’, namely ‘insertMany_’
      In a stmt of a 'do' block:
        insertMany_
          $ map (uncurry $ recordCons parentId)
              $ Set.toList (Set.difference newSet oldSet)
      In the second argument of ‘($)’, namely
        ‘do let oldSet = Set.fromList (zip ... old)
                newSet = Set.fromList (zip ... new)
            deleteWhere
              [parentFieldCons ==. parentId,
               indexFieldCons
                 <-. map fst (Set.toList $ Set.difference oldSet newSet)]
            insertMany_
              $ map (uncurry $ recordCons parentId)
                  $ Set.toList (Set.difference newSet oldSet)’
   |
71 |         insertMany_ $
   |         ^^^^^^^^^^^
~~~

--- a/cabal.config
+++ b/cabal.config
@@ -125,7 +125,7 @@ constraints:
   , path-io ==1.7.0
   , path-pieces ==0.2.1
   , pem ==0.2.4
-  , persistent ==2.13.3.5
+  , persistent ==2.14.4.3
   , persistent-sqlite ==2.13.1.0
   , persistent-template ==2.12.0.0
   , pretty ==1.1.3.6

--- a/package.yaml
+++ b/package.yaml
@@ -102,7 +102,7 @@ dependencies:
 # In order for Cabal (the tool) to build Stack, it needs to be told of the
 # upper bound on persistent. See
 # https://github.com/commercialhaskell/stack/pull/5677#issuecomment-1193318542
-- persistent >= 2.13.3.5 && < 2.14
+- persistent >= 2.14.0.0 && < 2.15
 - persistent-sqlite
 - persistent-template
 - pretty

--- a/src/Stack/Storage/Util.hs
+++ b/src/Stack/Storage/Util.hs
@@ -14,6 +14,10 @@ module Stack.Storage.Util
 
 import qualified Data.Set as Set
 import           Database.Persist
+                   ( BaseBackend, EntityField, PersistEntity
+                   , PersistEntityBackend, PersistField, PersistQueryWrite
+                   , SafeToInsert, (<-.), (==.), deleteWhere, insertMany_
+                   )
 import           Stack.Prelude
 import           Stack.Types.Storage ( StoragePrettyException (..) )
 
@@ -26,6 +30,7 @@ updateSet ::
        , PersistEntity record
        , MonadIO m
        , PersistQueryWrite backend
+       , SafeToInsert record
        )
     => (parentid -> value -> record)
     -> EntityField record parentid
@@ -51,6 +56,7 @@ updateList ::
        , PersistEntity record
        , MonadIO m
        , PersistQueryWrite backend
+       , SafeToInsert record
        )
     => (parentid -> Int -> value -> record)
     -> EntityField record parentid

--- a/stack-ghc-9.4.3.yaml
+++ b/stack-ghc-9.4.3.yaml
@@ -7,12 +7,6 @@ resolver: nightly-2022-12-22
 packages:
 - .
 
-extra-deps:
-# Stack will not build (at least, on Windows) with persistent > 2.13
-- persistent-2.13.3.5@sha256:adbcf7ac015722c1ada5e0899a5f012f459467302d148d903350fdab15bfd14f,7271
-
-allow-newer: true
-
 drop-packages:
 # See https://github.com/commercialhaskell/stack/pull/4712
 - cabal-install

--- a/stack-ghc-9.4.3.yaml.lock
+++ b/stack-ghc-9.4.3.yaml.lock
@@ -3,14 +3,7 @@
 # For more information, please see the documentation at:
 #   https://docs.haskellstack.org/en/stable/lock_files
 
-packages:
-- completed:
-    hackage: persistent-2.13.3.5@sha256:adbcf7ac015722c1ada5e0899a5f012f459467302d148d903350fdab15bfd14f,7271
-    pantry-tree:
-      sha256: 335542f76171368751a4a29846efdc642317ab85659ef38e6689ff90009a4a0c
-      size: 6427
-  original:
-    hackage: persistent-2.13.3.5@sha256:adbcf7ac015722c1ada5e0899a5f012f459467302d148d903350fdab15bfd14f,7271
+packages: []
 snapshots:
 - completed:
     sha256: 123d150c51c680f70864cb020315927fa4607f418d3dcccf25fd3a6263dc6717

--- a/stack.cabal
+++ b/stack.cabal
@@ -316,7 +316,7 @@ library
     , pantry >=0.8.1
     , path
     , path-io
-    , persistent >=2.13.3.5 && <2.14
+    , persistent >=2.14.0.0 && <2.15
     , persistent-sqlite
     , persistent-template
     , pretty
@@ -440,7 +440,7 @@ executable stack
     , pantry >=0.8.1
     , path
     , path-io
-    , persistent >=2.13.3.5 && <2.14
+    , persistent >=2.14.0.0 && <2.15
     , persistent-sqlite
     , persistent-template
     , pretty
@@ -565,7 +565,7 @@ executable stack-integration-test
     , pantry >=0.8.1
     , path
     , path-io
-    , persistent >=2.13.3.5 && <2.14
+    , persistent >=2.14.0.0 && <2.15
     , persistent-sqlite
     , persistent-template
     , pretty
@@ -695,7 +695,7 @@ test-suite stack-test
     , pantry >=0.8.1
     , path
     , path-io
-    , persistent >=2.13.3.5 && <2.14
+    , persistent >=2.14.0.0 && <2.15
     , persistent-sqlite
     , persistent-template
     , pretty

--- a/stack.yaml
+++ b/stack.yaml
@@ -5,10 +5,11 @@ packages:
 - .
 
 extra-deps:
+- fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
 - hpack-0.35.1@sha256:ef816234cbc7b52b0a6c55f7e904b6bc5292b8dd8f2d81ffcbcbc69ab80d75e5,4762
 - pantry-0.8.1@sha256:196111414d2489499fda6213deebcb865bc12285023d5af9bd273bf27cdb185d,4099
+- persistent-2.14.4.3@sha256:22237669897fed2b71de521bfeba7f8e0677e54992920288ae1fa8206a4711fc,7199
 - rio-prettyprint-0.1.4.0@sha256:1f8eb3ead0ef33d3736d53e1de5e9b2c91a0c207cdca23321bd74c401e85f23a,1301
-- fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
 
 drop-packages:
 # See https://github.com/commercialhaskell/stack/pull/4712

--- a/stack.yaml.lock
+++ b/stack.yaml.lock
@@ -5,6 +5,13 @@
 
 packages:
 - completed:
+    hackage: fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
+    pantry-tree:
+      sha256: f8e89e81fe0ca6e285036e04b165119ccfc59e7e668c1bfc28927f39062c5949
+      size: 1280
+  original:
+    hackage: fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
+- completed:
     hackage: hpack-0.35.1@sha256:ef816234cbc7b52b0a6c55f7e904b6bc5292b8dd8f2d81ffcbcbc69ab80d75e5,4762
     pantry-tree:
       sha256: 514f0311eff442ab83ef3c2db456857f360ea173c17df133e575f50915e02da0
@@ -19,19 +26,19 @@ packages:
   original:
     hackage: pantry-0.8.1@sha256:196111414d2489499fda6213deebcb865bc12285023d5af9bd273bf27cdb185d,4099
 - completed:
+    hackage: persistent-2.14.4.3@sha256:22237669897fed2b71de521bfeba7f8e0677e54992920288ae1fa8206a4711fc,7199
+    pantry-tree:
+      sha256: 757771197e4ede18c33addcb4c5b6cf7cd38511b2d9f90e73f1fb41bb482f2ae
+      size: 6601
+  original:
+    hackage: persistent-2.14.4.3@sha256:22237669897fed2b71de521bfeba7f8e0677e54992920288ae1fa8206a4711fc,7199
+- completed:
     hackage: rio-prettyprint-0.1.4.0@sha256:1f8eb3ead0ef33d3736d53e1de5e9b2c91a0c207cdca23321bd74c401e85f23a,1301
     pantry-tree:
       sha256: 05bb2585e2d93899c62e0b6e1da02d040286b4524d0d363be1c5d99e49ac95bd
       size: 628
   original:
     hackage: rio-prettyprint-0.1.4.0@sha256:1f8eb3ead0ef33d3736d53e1de5e9b2c91a0c207cdca23321bd74c401e85f23a,1301
-- completed:
-    hackage: fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
-    pantry-tree:
-      sha256: f8e89e81fe0ca6e285036e04b165119ccfc59e7e668c1bfc28927f39062c5949
-      size: 1280
-  original:
-    hackage: fsnotify-0.4.1.0@sha256:44540beabea36aeeef930aa4d5f28091d431904bc9923b6ac4d358831c651235,2854
 snapshots:
 - completed:
     sha256: a2cbcd2f37010a64c4ef74c21fd7e55982a07b49840d2bed306f9bac9981a9c3

